src/config.rs

```diff
@@ -392,33 +413,59 @@ impl quilkin_xds::config::Configuration for Config {
         tracing::trace!("waiting for changes");
 
         async move {
-            let mut cluster_watcher = control_plane.config.clusters.watch();
-            match &control_plane.config.datacenter {
-                crate::config::DatacenterConfig::Agent { .. } => loop {
-                    match cluster_watcher.changed().await {
-                        Ok(()) => control_plane.push_update(crate::xds::CLUSTER_TYPE),
-                        Err(error) => tracing::error!(%error, "error watching changes"),
-                    }
-                },
+            let clusters = control_plane.config.dyn_cfg.clusters();
+
+            let dc = match &control_plane.config.datacenter {
+                crate::config::DatacenterConfig::Agent { .. } => None,
                 crate::config::DatacenterConfig::NonAgent { datacenters } => {
-                    let mut dc_watcher = datacenters.watch();
+                    Some(datacenters.clone())
+                }
+            };
+
+            match (clusters, dc) {
+                (Some(clusters), Some(dc)) => {
```

Comment:
As a followup, I think it would be good if this implemented a little differently, where we iterate over the typemap, and get the watcher for each type, and when any of those watchers changes it pushes an update with it's matching xds type.

The typemap library we're using doesn't have a `.iter()`, but the type is just a wrapper around `HashMap`, so it should be straightforward to add that.